testcases/cve/stack_clash.c

```diff
@@ -0,0 +1,267 @@
+ * after `mmap` are moving stack further.
+ *
+ * [1] https://blog.qualys.com/securitylabs/2017/06/19/the-stack-clash
+ * [2] https://bugzilla.novell.com/show_bug.cgi?id=CVE-2017-1000364
+ */
+
+#include <sys/mman.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <alloca.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include "tst_test.h"
+#include "tst_safe_stdio.h"
+
+static unsigned long PAGE_SIZE;
+static unsigned long PAGE_MASK;
+static unsigned long GAP_PAGES = 256;
+static unsigned long THRESHOLD;
+static int STACK_GROWSDOWN;
+
+#define SIGNAL_STACK_SIZE	(1UL<<20)
+#define FRAME_SIZE		1024
+#define REQ_STACK_SIZE		(1024 * 1024)
+
+#define EXIT_TESTBROKE		TBROK
+
+void exhaust_stack_into_sigsegv(void)
+{
+	volatile char * ptr = alloca(FRAME_SIZE - sizeof(long));
+	*ptr = '\0';
+	exhaust_stack_into_sigsegv();
+}
+
+#define MAPPED_LEN PAGE_SIZE
+static unsigned long mapped_addr;
+
+void segv_handler(int sig, siginfo_t *info, void *data LTP_ATTRIBUTE_UNUSED)
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV && sig != SIGBUS)
+		return;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO,
+		"mmap = [%lx, %lx), addr = %lx, diff = %lx, THRESHOLD = %lx",
+		mapped_addr, mmap_end, fault_addr, diff, THRESHOLD);
+	if (diff < 0 || (unsigned long)diff < THRESHOLD)
+		_exit(EXIT_FAILURE);
+	else
+		_exit(EXIT_SUCCESS);
+}
+
+unsigned long read_stack_addr_from_proc(unsigned long *stack_size)
+{
+	FILE *fh;
+	char buf[1024];
+	unsigned long stack_top = -1UL, start, end;
+
+	fh = SAFE_FOPEN("/proc/self/maps", "r");
+
+	while (!feof(fh)) {
+		if (fgets(buf, sizeof(buf), fh) == NULL) {
+			tst_brk(TBROK | TERRNO, "fgets");
+			goto out;
+		}
+
+		if (!strstr(buf, "[stack"))
+			continue;
+
+		if (sscanf(buf, "%lx-%lx", &start, &end) != 2) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			goto out;
+		}
+
+		*stack_size = end - start;
+
+		if (STACK_GROWSDOWN)
+			stack_top = start;
+		else
+			stack_top = end;
+		break;
+	}
+
+out:
+	SAFE_FCLOSE(fh);
+	return stack_top;
+}
+
+void dump_proc_self_maps(void)
+{
+	static char buf[64];
+	static const char *cmd[] = {"cat", buf, NULL};
+	sprintf(buf, "/proc/%d/maps", getpid());
+	tst_run_cmd(cmd, NULL, NULL, 0);
+}
+
+void preallocate_stack(unsigned long required)
+{
+	volatile char *garbage;
+
+	garbage = alloca(required);
+	garbage[0] = garbage[required - 1] = '\0';
+}
+
+void do_child(void)
+{
+	unsigned long stack_addr, stack_size;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	void *map;
+	unsigned long gap = GAP_PAGES * PAGE_SIZE;
+	struct rlimit rlimit;
+
+	rlimit.rlim_cur = rlimit.rlim_max = RLIM_INFINITY;
+	SAFE_SETRLIMIT(RLIMIT_STACK, &rlimit);
+
+	preallocate_stack(REQ_STACK_SIZE);
```

Comment:
> It looks like sometimes a call to mmap or sigaction uses a bit more of a stack, moving the top too close

I think what you are saying is that stack can grow  between calls to read_stack_addr_from_proc() and mmap(), and we end up mapping right next to stack, which the kernel won't treat as violation of guard gap.

Calls expanding stack after mmap/signal should be free to try that - kernel is responsible for catching any violations of guard gap.

"preallocate_stack" makes sense to me now, 1M seems a bit too much, but it shouldn't hurt anything.